blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md

```diff
@@ -0,0 +1,116 @@
+---
+title: Porting EDK II to an old Allwinner A13 tablet
+abstract: 'An effort to port EDK II to Allwinner SoCs'
+cover: /covers/image-file.png
+author: artur.kowalski
+layout: post
+published: true
+date: 2021-10-05
+archives: "2021"
+
+tags:
+  - firmware
+categories:
+  - Firmware
+  - Manufacturing
+
+---
+
+# Introduction
+
+I have an old cheap Allwinner A13 tablet with low memory and a weak CPU; it's
+slow even for internet browsing, so I've been using it primarily for
+development. Some time ago, I started porting EDK II to it. I plan to replace
+the current Allwinner bootloader with a full-featured UEFI implementation
+capable of booting mainline Linux, FreeBSD, and possibly Windows 10, all running
+with ACPI. Using ACPI would allow OS to boot without SoC specific drivers, as
+all basic functionality like CPU/RAM detection, frequency, and voltage scaling,
+serial port, RTC, etc., would be implemented by ACPI and runtime drivers.
+
+## Project status
+
+The project is still at its early stage - it does not support CPU frequency
+scaling, so it's running very slow. Currently, there is basic support for some
+hardware: UART, I2C, PMIC, SD (eMMC support is broken).
+
+It is possible to boot recent Linux with FDT (tested with v5.11).
+
+Following things need to be implemented
+- CPU frequency scaling
+- LCD
+- HDMI
+- DMA
+- UEFI variables - currently, variables are not preserved across reboots
+- ACPI
+- Support for other CPUs and boards.
+
+Currently, I am working on USB peripheral mode support. It will allow making
+boot process compliant with Pi specification. Currently, a small executable
+(called SPL after U-Boot's SPL) is loaded into the device's memory, it
+initializes DRAM and returns into BootROM, which proceeds into loading UEFI
+image. Implementing USB driver makes it possible to load EFI directly from Pi
+without going back into BootROM, thus allowing to move initialization code
+there. Implementing it also enables UEFI to be debugged using USB instead of
+UART.
+
+## Supported hardware
+
+Currently, only my XW711 tablet is supported, but this hardware is so similar
+to Q8 that it should work on any Q8 tablet, these look like the same board, with
+only some peripherals like touch panel and LCD differing.
+
+Since there is no display support and USB is still on its way, UART connection
+is required, on this board (and on Q8), UART pads are located on the back of the
+board, there is also a second UART multiplexed with µSD, which can be used
+without disassembling device, support for this is coming soon.
+
+## Building and booting UEFI
+
+Source code is available
+[here](https://github.com/arturkow2000/SunxiPlatformPkg), you can build it using
+`arm-none-eabi-gcc`. 
```

Comment:
@arturkow2000 where to get it and what version was tested?

Reply:
maybe using 3mdeb docker for edk2 would be more scalable and less dependent on readers OS